crates/red_knot_python_semantic/src/types/property_tests/type_bound.rs

```diff
@@ -0,0 +1,279 @@
+                                match (pos.len(), neg.len()) {
+                                    // an empty intersection does not mean
+                                    // anything
+                                    (0, 0) => None,
+                                    // a single positive element should be
+                                    // unwrapped
+                                    (1, 0) => Some(pos[0].clone()),
+                                    _ => Some(ty),
+                                }
+                            } else {
+                                unreachable!()
+                            }
+                        }),
+                )
+            }
+            _ => Box::new(std::iter::empty()),
+        }
+    }
+}
+
+static TYPE_POOL: &[Ty] = &[
+    Ty::Never,
+    Ty::Unknown,
+    Ty::None,
+    Ty::Any,
+    Ty::IntLiteral(0),
+    Ty::IntLiteral(1),
+    Ty::IntLiteral(-1),
+    Ty::BooleanLiteral(true),
+    Ty::BooleanLiteral(false),
+    Ty::StringLiteral(""),
+    Ty::StringLiteral("a"),
+    Ty::LiteralString,
+    Ty::BytesLiteral(""),
+    Ty::BytesLiteral("\x00"),
+    Ty::KnownClassInstance(KnownClass::Object),
+    Ty::KnownClassInstance(KnownClass::Str),
+    Ty::KnownClassInstance(KnownClass::Int),
+    Ty::KnownClassInstance(KnownClass::Bool),
+    Ty::KnownClassInstance(KnownClass::List),
+    Ty::KnownClassInstance(KnownClass::Tuple),
+    Ty::KnownClassInstance(KnownClass::FunctionType),
+    Ty::KnownClassInstance(KnownClass::SpecialForm),
+    Ty::KnownClassInstance(KnownClass::TypeVar),
+    Ty::KnownClassInstance(KnownClass::TypeAliasType),
+    Ty::KnownClassInstance(KnownClass::NoDefaultType),
+    Ty::TypingLiteral,
+    Ty::BuiltinClassLiteral("str"),
+    Ty::BuiltinClassLiteral("int"),
+    Ty::BuiltinClassLiteral("bool"),
+    Ty::BuiltinClassLiteral("object"),
+    Ty::BuiltinInstance("type"),
+    Ty::AbcInstance("ABC"),
+    Ty::AbcInstance("ABCMeta"),
+    Ty::SubclassOfAny,
+    Ty::SubclassOfBuiltinClass("object"),
+    Ty::SubclassOfBuiltinClass("str"),
+    Ty::SubclassOfBuiltinClass("type"),
+    Ty::AbcClassLiteral("ABC"),
+    Ty::AbcClassLiteral("ABCMeta"),
+    Ty::SubclassOfAbcClass("ABC"),
+    Ty::SubclassOfAbcClass("ABCMeta"),
+    Ty::AlwaysTruthy,
+    Ty::AlwaysFalsy,
+    Ty::BuiltinsFunction("chr"),
+    Ty::BuiltinsFunction("ascii"),
+    Ty::BuiltinsBoundMethod {
+        class: "str",
+        method: "isascii",
+    },
+    Ty::BuiltinsBoundMethod {
+        class: "int",
+        method: "bit_length",
+    },
+];
+
+fn filter_type_pool(predicate: fn(db: &TestDb, ty: &Ty) -> bool) -> &'static [Ty] {
+    let db = get_cached_db();
+    let pool = TYPE_POOL
+        .iter()
+        .filter(|ty| predicate(&db, ty))
+        .cloned()
+        .collect::<Vec<Ty>>();
+
+    assert!(!pool.is_empty(), "No types found in the pool");
+
+    Box::leak(pool.into_boxed_slice())
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct AnyTy;
```

Comment:
This represents an arbitrary gradual type. Should we maybe rename it to `GradualType`/`GradualTy` to avoid any confusion with `typing.Any`?